// @author.io/dev/commands/workspace.js
import { Command } from '@author.io/node-shell'
import { container, launch, abort, hr } from '../lib/utility.js'
import flags from '../lib/flags.js'

// TODO: Make an auto-rebuild feature for file changes.
const WorkspaceCommand = new Command({
  name: 'workspace',
  alias: 'ws',
  description: 'Launch a terminal within a Linux container. The local project is loaded in a working directory named /app',
  flags: {
    runtime: flags.runtime
  },
  async handler (meta) {
    const { version, image, command } = await container(meta).catch(abort)

    if (meta.flags.unrecognized.length > 0) {
      command.push(meta.flags.unrecognized.join(' '))
    }

    command.push(`${image}${version ? ':' + version : ''} bash`)

    if (!meta.flag('verbose')) {
      console.clear()
      console.log(hr(40))
      console.log('Running in ' + image)
      console.log('Type "exit" to quit the terminal.')
      console.log(hr(40))
    }

    launch(command, meta.flag('verbose'))
  }
})

export { WorkspaceCommand as default }
